### ISSUE TITLE
Carrying element context through streams

### ISSUE THREAD

**Issue by jroper** [akka/akka#25394]
This issue is intended as a discussion about whether it makes sense to have a per element context, and how that might be implemented. I don't know if it would be a good idea to provide this or not, but I was thinking about it while riding to work today so I thought I'd record it somewhere.

Consider Kafka streams. I might be doing a stream transformation that looks like this:

```java
KStream<String, Message> stream = ...;
KStream<String, Transformed> transformed = stream
  .filter((key, message) -> message.isInteresting())
  .map((key, message) -> KeyValue.pair(key, transform(message)));
```

Now the important thing to realise here is that when the transformed messages are emitted, they are associated with the original message that came in, and this is so that once the transformed messages are emitted, the original message that caused it can be acknowledged, giving at least once delivery guarantees. This only works because Kafka streams tracks which incoming message is associated with each outgoing message - it carries this context through the stream processing. In Akka streams we can't do this, we only have two options, either require users to explicitly acknowledge errors (which in practice is error prone because you can't for example handle and then acknowledge a message in the same mapAsync since it could allow an acknowledgement of a later message to overtake the failure of an earlier message), or we can offer an API that allows you to do 1:1 transformations on messages, and carry that context around the transformation making the assumption that we can zip it back together because it's a 1:1 transformation, but as soon as the user does a transformation that isn't 1:1 like `filter` or `mapConcat`, it breaks down. We already know from experience offering APIs like this in Lagom that this causes confusion and bugs for users.

So the question is, can we implement some form of message context, which would then allow us to offer streaming APIs for message broker transformation that are as simple and intuitive to use as Kafka Streams? I think we can.

Now of course, there are times when it's impossible to carry this context. For example, the moment you insert a reactive streams processor into the flow, you're going to lose the context. There's nothing we can do about that. But, from code that I've seen out in the wild, I've actually never even seen someone do that. It's all `map`, `mapAsync`, `filter`, `grouped`, `mapConcat`, all operations that can carry context (`grouped` is an interesting one that I'll get to later). So, I think it's fine to have such a limitation, and to just allow the context to be dropped in certain situations.

So let's take a first shot at how such an API might look:

```scala
trait FlowOps[In] {
  def viaWithContext[Out](flow: Flow[In, Out, _]): Flow[In, (Option[In], Out), NotUsed]
}
```

This method would be used by a library, that took a user supplied `flow`, and injected it into its stream. For each element emitted by the user supplied flow, it will then, if it can determine the incoming message context, associate it with the output message, and emit a tuple of the two. How might this be implemented in practice? Considering a graph stage logic, whenever an `onPush` callback was invoked, the current context message (if such context exists) that caused the push would be set as the context for that stage, and each `push` that that stage did from then until the next `onPush` was invoked would have the current context message for that stage set as the context for the next down stream `onPush`. For simple stages like `map`/`filter`, this will just work. For more complex stages, such as `mapAsync`, the stage will need to be aware of context - it will need to capture the current context when it dispatches the processing, and then re-associate the captured context for that operation with the stage when it handles the asynchronous callbacks. For something like `grouped`, it means that only the context that triggered the group to be emitted will be carried as the context - for the Kafka streams use case, this is exactly what we want, as when we group, we will end up only wanting to commit the last message in the group. For merges (eg `flatMapConcat` or `flatMapMerge`), each input context will need to be associated with its corresponding sub streams.

Nested `viaWithContext`'s would be a challenge, we could probably just disallow them.

Now there is one problem with the above in practice when applied to our use case of providing a Kafka Streams like API - that is, when a stage emits multiple elements per input element, you only want to acknowledge the incoming element for the last element associated with that incoming element from the stage, but there's no way to know when you receive an element whether any further elements are going to be received for that context. One possible solution to this might be to allow stages to signal, when they push, whether the this terminates the incoming context or not. They would only signal termination if the incoming context had the termination signal attached to it. Sometimes, there might be a message to go with the terminating push (eg, 1:1 stages like `map`, stages where all output elements are known like `mapConcat`), but at other times, like `flatMapConcat`, this signal will have to be pushed without a corresponding output element. Also, a stage like `filter` will need to push the terminate signal if it filters out a message when the current incoming context has the terminate signal associated with it, otherwise the terminate signal will be dropped. An API for this might look like this:

```scala
sealed trait Context[+T]
case object NoContext extends Context[Nothing]
case class TerminateContext[+T](element: T) extends Context[T]
case class ContinueContext[+T](element: T) extends Context[T]

trait FlowOps[In] {
  def viaWithContext[Out](flow: Flow[In, Out, _]): Flow[In, (Context[In], Option[Out]), NotUsed]
}
```

Now all of this is very complex, there is another completely different possibility for how this might be implemented underneath, and that is to use a different materializer for the nested stages. I don't understand the materialization process of Akka streams well enough to be able to know how feasible this is, but the idea would be essentially to walk through the graph, and manually transform each stage so that it carries context. Support would need to be implemented specific to each stage, but an advantage here is that it could fail if there was a stage that didn't support carrying context (and in this case, no custom graph stage would support carrying context). So a transformation for `Map` might look like this:

```scala
case Map(f) => Map {
  case (ctx, None) => (ctx, None)
  case (ctx, Some(t)) => (ctx, Some(f(t))))
}
```

While a transformation of `Filter` might look like this:

```scala
case Filter(predicate) => Collect {
  case (ctx, None) => (ctx, None)
  case (ctx: TerminateContext, t) => (ctx, t.filter(predicate))
  case (ctx, o @ Some(t)) if predicate(t) => (ctx, o)
}
```

A hybrid of the two approaches could also be done where graph stages could declare a `viaWithContext` version of themselves.

**Comment by johanandren** [comment]
I think this is a duplicate of, or at least closely related to #15957

We had one idea recently (as in the last year or so) that may be interesting. The idea was to provide the "focusing" as an interface like `SubStream` that does only support a subset of the capabilities of streams and does not allow any operations that would drop elements or re-order them. I think this could be done without touching the materializer or running stream internals.

**Comment by ktoso** [comment]
Similar things were discussed over the years:
- https://github.com/akka/akka-stream-contrib/pull/52/files  talpr-50-flow-tagger  "focus, unfocus"
- https://github.com/akka/akka/issues/22975 Flow tuple projections
- and even https://github.com/akka/akka/issues/15957 Add new graph element: `contextBypass`

May I request you look into those and reconciliate what the requirement being stated is? Which of these shall we close.

**Comment by jrudolph** [comment]
By now it's pretty clear that such a functionality would be useful in many cases. On the other hand it's also pretty clear that a general solution is impossible or very hard to do requiring support in every single stage.

The problem I see with attempting a general solution is that apart from the few absolutely clear cases, many cases may introduce another configuration option for the stage because there might not be the one clear-cut way to attach the context. `grouped` would be one such example where taking the last context makes sense for Kafka but might be the wrong choice for other use cases.

For me that sounds like we shouldn't attempt a general solution because the trade-off between general benefit and overall complexity doesn't seem favorable for adding this functionality in the core (i.e. GraphStage).

When we talked about it we thought about a `SubStream`-like approach, i.e. we could try to invent a subset API that works exactly for the cases that we know would be useful for Kafka. The `viaWithContext` API you suggest seems to be similar but wouldn't prevent building "illegal" flows but would fail at runtime when a component of the flow cannot carry the context.

E.g. like this:

```scala
Flow[In]
  .withContext
     .map(...)
     .filter(...)
     .mapAsync(...)
  .provideContext // (Out, In)
```

The good thing would be that it could be implemented in a quite non-intrusive way on the side.

**Comment by jrudolph** [comment]
Here is a simple implementation: https://github.com/jrudolph/akka/commit/5b9454dd76f93dac72668bddf320812d4c7913e5

(though not tested at all).

**Comment by 2m** [comment]
In Alpakka Kafka we have a couple of use-cases where we use context (there it is called PassThrough). All of these are implemented for that particular use-case and without generalizations outside of Alpakka Kafka:

* consuming one message, producing more than one, and finally committing the consumed offset. Offset is held in PassThrough. https://github.com/akka/alpakka-kafka/blob/cb3ff9d910c24629f9b4d52e37e9516ca852f14f/tests/src/test/scala/docs/scaladsl/AtLeastOnce.scala#L38-L56
* filtering out messages, but still committing offsets later https://github.com/akka/alpakka-kafka/blob/cb3ff9d910c24629f9b4d52e37e9516ca852f14f/tests/src/test/scala/docs/scaladsl/AtLeastOnce.scala#L91-L92

**Comment by adkafka** [comment]
What do we think about adding a new flow stage: `MapRightUnordered`, that is implemented as a BidiFlow. The job of `MapRightUnordered` is to pass the Right element of a 2-tuple through a given flow stage. The output from the flow stage will then be emitted as the Right element of a 2-tuple. The Left element of the tuple will be a `PassThrough` element. This requires some kind of identifier to retrieve a `PassThrough` key from an output element.

I think this interface would work well for the end user:
```
  /**
   * Provides a Pass Through abstraction over an inner flow, by tracking which elements are "in flight" (Input element
   * waiting for a corresponding Output element), and storing a [[PassThrough]] in the flow stage via a map. When an
   * Output element is returned from the Inner flow, the corresponding [[PassThrough]] element will be extracted (using
   * the key returned from `outputExtractor` ), and emitted from this flow stage. This means that an identifier must be
   * present on the [[Out]] type that can be translated into a Key known during Input (transaction ID, UUID, etc).
   *
   * Note: These possible exceptions can be thrown, and must be handled via the stream supervisor:
   * * [[DequeueException]] - [[Out]] yielded a key [[K]] that has not been encountered before
   * * [[QueueTimeoutException]] - An element has remained in the queue for longer than the configured time out
   *
   * '''Emits when''' The inner flow stage emits
   *
   * '''Backpressures when''' the number of futures reaches the configured parallelism and the downstream backpressures
   *
   * '''Completes when''' upstream completes and there are 0 in flight messages
   *
   * '''Cancels when''' inner stream or downstream cancels
   *
   * @param parallelism     Limit the size of the map that tracks "in-flight" requests. This has the affect of applying
   *                        backpressure, in the same way as `mapAsync` and similar flow stages.
   * @param innerFlow       Flow stage to wrap. If this flow stage drops elements silently, this will result in
   *                        QueueTimeoutExceptions and will negatively affect performance. The resulting
   *                        [[MapRightUnordered]] will have the same ordering guarantees as the innerFlow that it wraps.
   * @param outputExtractor Function for extracting a [[PassThrough]] Key from the [[Out]] type
   * @param elementTimeout  How long an element can remain "in-flight" before a [[QueueTimeoutException]] is thrown. This
   *                        timeout is checked on regular intervals.
   * @tparam In Input type of the inner flow stage
   * @tparam Out Output type of the inner flow stage
   * @tparam K Key type of the PassThrough
   * @tparam V Value type of the PassThrough
   * @tparam M Materialized type from inner flow stage
   * @return Flow stage that maps over the right element of a 2-tuple
   */
  def apply[In, Out, K, V, M](parallelism: Int)
    (innerFlow: Flow[In, Out, M])
    (outputExtractor: Out => K)
    (implicit elementTimeout: FiniteDuration = 5.seconds): Graph[FlowShape[(PassThrough[K, V], In), (PassThrough[K, V], Out)], M
```

Where `PassThrough` is defined as such:
```
  trait PassThrough[K, V] {
    def key: K
    def value: V
  }
```

I have already implemented a prototype for this interface, so I would be happy to PR it if it seems fit. I think we could provide `MapRight` as well, which will provide ordering guarentee (similar to `MapAsync` vs `MapAsyncUnordered`).

**Comment by jrudolph** [comment]
Let's close this one as implemented in #25951.
